src/Substance.hs

```diff
@@ -441,6 +457,50 @@ substInsert sigma y arg = case M.lookup y sigma of
 --                               Nothing -> Right "Argument " ++ (show v) ++ " not in environment\n"
 --                               vType -> Left vType
 
+----------------------------------------- Binding & Equality Environment ------------------
+-- | Definition of the suBSTANCE environemt + helper functions.
+--   Contains binding information and equality of expressions and predicates
+--   In order to calculate all the equalities, we cmpute the closure of the
+--   equlities in Substance + symetry
+
+data SubEnv = SubEnv { exprEqualities :: [(Expr,Expr)],
+                       predEqualities :: [(Predicate,Predicate)],
+                       bindings       :: M.Map Var Expr,
+                       subPreds :: [Predicate]
+                    }
+                     deriving (Show, Eq, Typeable)
+
+-- | The top levek function for computing the Substance environement
+--   Important: this function assuemes it runs after the typechecker and that
+--              the program is well-formed (as well as the DSLL)
+loadSubEnv :: SubProg -> SubEnv
+loadSubEnv p = let subEnv1 = foldl loadStatements initE p
+                   subEnv2 = computeEqualityClosure subEnv1
+               in subEnv2
+              where initE = SubEnv {exprEqualities = [], predEqualities = [], bindings = M.empty, subPreds = []}
+
+loadStatements :: SubEnv -> SubStmt -> SubEnv
```

Comment:
this is better named as `loadStatement`

Reply:
you might want to comment that the order in all the lists is reversed (in case anyone is relying on order of Substance statements... hopefully not)